python/cudf_polars/cudf_polars/dsl/utils/windows.py

```diff
@@ -0,0 +1,197 @@
+
+if TYPE_CHECKING:
+    from cudf_polars.typing import ClosedInterval, Duration
+
+
+__all__ = [
+    "duration_to_int",
+    "duration_to_scalar",
+    "offsets_to_windows",
+    "range_window_bounds",
+]
+
+
+def duration_to_int(
+    dtype: plc.DataType,
+    months: int,
+    weeks: int,
+    days: int,
+    nanoseconds: int,
+    parsed_int: bool,  # noqa: FBT001
+    negative: bool,  # noqa: FBT001
+) -> int:
+    """
+    Convert a polars duration value to an integer.
+
+    Parameters
+    ----------
+    dtype
+        The type of the column being added to.
+    months
+        Number of months
+    weeks
+        Number of weeks
+    days
+        Number of days
+    nanoseconds
+        Number of nanoseconds
+    parsed_int
+        Is this actually a representation of an integer, not a duration?
+    negative
+        Is this a negative duration?
+
+    Returns
+    -------
+    Pyarrow scalar
+        With datatype matching the provided dtype.
+
+    Raises
+    ------
+    NotImplementedError
+        For unsupported durations or datatypes.
+    """
+    if months != 0:
+        raise NotImplementedError("Month durations in rolling windows")
+    if parsed_int and (weeks != 0 or days != 0 or dtype.id() != plc.TypeId.INT64):
+        raise NotImplementedError(
+            "Invalid duration for parsed_int"
+        )  # pragma: no cover; polars raises first
+    elif not parsed_int and dtype.id() == plc.TypeId.INT64:
+        raise pl.exceptions.InvalidOperationError("Duration must be a parsed integer")
+    value = nanoseconds + 24 * 60 * 60 * 10**9 * (days + 7 * weeks)
+    return -value if negative else value
+
+
+def duration_to_scalar(dtype: plc.DataType, value: int) -> pa.Scalar:
+    """
+    Convert a raw polars duration value to a pyarrow scalar.
+
+    Parameters
+    ----------
+    dtype
+        The type of the column being added to.
+    value
+        The raw value as in integer. If `dtype` represents a timestamp
+        type, this should be in nanoseconds.
+    months
+        Number of months
+    weeks
+        Number of weeks
+    days
+        Number of days
+    nanoseconds
+        Number of nanoseconds
+    parsed_int
+        Is this actually a representation of an integer, not a duration?
+    negative
+        Is this a negative duration?
+
+    Returns
+    -------
+    Pyarrow scalar
+        With datatype matching the provided dtype.
+
+    Raises
+    ------
+    NotImplementedError
+        For unsupported durations or datatypes.
+    """
+    tid = dtype.id()
+    if tid == plc.TypeId.INT64:
+        return pa.scalar(value, type=pa.int64())
+    elif tid == plc.TypeId.TIMESTAMP_NANOSECONDS:
+        return pa.scalar(value, type=pa.duration("ns"))
+    elif tid == plc.TypeId.TIMESTAMP_MICROSECONDS:
+        return pa.scalar(value // 10**3, type=pa.duration("us"))
+    elif tid == plc.TypeId.TIMESTAMP_MILLISECONDS:
+        return pa.scalar(value // 10**6, type=pa.duration("ms"))
```

Comment:
If possible could we construct `pylibcudf.Scalar`s here with `from_py`? It would help with the pyarrow-optional-dependency effort https://github.com/rapidsai/cudf/issues/17046

Reply:
I can't do this because I cannot produce nanosecond precision scalars with any of the `from_py` interfaces

Reply:
I think https://github.com/rapidsai/cudf/pull/18563 should provide support for this, so I can address this later

Reply:
Ah, thanks!